
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Compare nanopore signal to find modifications">
      
      
        <meta name="author" content="Adrien Leger & Tommaso Leonardi">
      
      
        <link rel="canonical" href="http://nanocompore.rna.rocks/output/">
      
      
        <link rel="prev" href="../usage/">
      
      
        <link rel="next" href="../plotting/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Outputs - Nanocompore</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
      


    
    
      
    
    
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#outputs" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Nanocompore" class="md-header__button md-logo" aria-label="Nanocompore" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Nanocompore
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Outputs
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/tleonardi/nanocompore" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    tleonardi/nanocompore
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Nanocompore" class="md-nav__button md-logo" aria-label="Nanocompore" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Nanocompore
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/tleonardi/nanocompore" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    tleonardi/nanocompore
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation instructions
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../data_preparation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data preparation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running Nanocompore
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Outputs
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Outputs
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#results-tsv" class="md-nav__link">
    <span class="md-ellipsis">
      Results TSV
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Results TSV">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mandatory-output-columns" class="md-nav__link">
    <span class="md-ellipsis">
      Mandatory output columns
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gaussian-mixture-models-gmm-output-columns" class="md-nav__link">
    <span class="md-ellipsis">
      Gaussian-Mixture Models (GMM) output columns
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#shift-statistics-tsv" class="md-nav__link">
    <span class="md-ellipsis">
      Shift statistics TSV
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#result-database" class="md-nav__link">
    <span class="md-ellipsis">
      Result database
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Plotting
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data_api/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data API
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting_api/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Plotting API
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../troubleshooting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Troubleshooting
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../alternative/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Alternative and complementary packages
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../citing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../code_of_conduct/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Code of conduct
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../licence/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Licence
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../changelog/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Changelog
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#results-tsv" class="md-nav__link">
    <span class="md-ellipsis">
      Results TSV
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Results TSV">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mandatory-output-columns" class="md-nav__link">
    <span class="md-ellipsis">
      Mandatory output columns
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gaussian-mixture-models-gmm-output-columns" class="md-nav__link">
    <span class="md-ellipsis">
      Gaussian-Mixture Models (GMM) output columns
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#shift-statistics-tsv" class="md-nav__link">
    <span class="md-ellipsis">
      Shift statistics TSV
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#result-database" class="md-nav__link">
    <span class="md-ellipsis">
      Result database
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="outputs">Outputs</h1>
<p>While running, Nanocompore stores the results in an SQLite database. After all transcripts are analyzed, it would perform a postprocessing step and output an easy to read TSV (tab-separated values) file.</p>
<h2 id="results-tsv">Results TSV</h2>
<p>The results TSV is found in the output directory, typically named as "out_nanocompore_results.tsv". There's a set of columns that always will be present in the results and others that may appear or not, depending on the input configuration. For example, changing the statistical tests that are performed by Nanocompore would change the set of output columns.</p>
<h3 id="mandatory-output-columns">Mandatory output columns</h3>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>pos</strong></td>
<td>Starting position on the transcript's reference sequence of the k-mer for which the row reports results. Indexing is 0-based, meaning that the first nucleotide on the transcript is position 0.</td>
<td>42</td>
</tr>
<tr>
<td><strong>chr</strong></td>
<td>Chromosome id. This will be set only if a GTF file is provided using the <code>gtf</code> parameter in the configuration.</td>
<td>chr1</td>
</tr>
<tr>
<td><strong>genomicPos</strong></td>
<td>The position on the chromosome corresponding to the position on the transcript. This is only set when <code>gtf</code> file is provided.</td>
<td>1324576</td>
</tr>
<tr>
<td><strong>ref_id</strong></td>
<td>Reference name of the transcript.</td>
<td>ENST00000676788.1</td>
</tr>
<tr>
<td><strong>strand</strong></td>
<td>Genomic strand. This is set only when <code>gtf</code> file is provided.</td>
<td>+</td>
</tr>
<tr>
<td><strong>ref_kmer</strong></td>
<td>The k-mer sequence as found in the reference.</td>
<td>GCTACGT</td>
</tr>
</tbody>
</table>
<h3 id="gaussian-mixture-models-gmm-output-columns">Gaussian-Mixture Models (GMM) output columns</h3>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>GMM_chi2_pvalue</strong></td>
<td>p-value obtained from performing an association testing with a Chi-squared test between the reads' condition labels and the cluster assignments obtained from the GMM.</td>
<td>0.00123</td>
</tr>
<tr>
<td><strong>GMM_chi2_qvalue</strong></td>
<td>The p-value corrected for multiple testing using the Benjamini-Hochberg procedure.</td>
<td>1.0</td>
</tr>
<tr>
<td><strong>GMM_LOR</strong></td>
<td>The log odds ratio from the GMM. Suppose that we have a wildtype (WT) condition and a mod-writer knockdown (KD) condition and that we have clusters C1 and C2 detected by the GMM, then <code>GMM_LOR=ln((WT_C1/WT_C2)/(KD_C1/KD_C2))</code>, where <code>ln</code> denotes the natural logarithm.</td>
<td>1.3</td>
</tr>
<tr>
<td><strong>&lt;SAMPLE&gt;_mod</strong></td>
<td>The number of reads for the given k-mer assigned to the GMM cluster considered to represent the modification state.</td>
<td>47</td>
</tr>
<tr>
<td><strong>&lt;SAMPLE&gt;_unmod</strong></td>
<td>The number of reads for the given k-mer assigned to the GMM cluster considered to represent the non-modification state.</td>
<td>72</td>
</tr>
</tbody>
</table>
<p>The last two columns are repeated for each of the samples provided in the input configuration YAML file. <code>&lt;SAMPLE&gt;</code> will be substituted with the sample label used in the configuration.</p>
<p><strong>IMPORTANT:</strong>
It's recommended that both the q-value and the <code>GMM_LOR</code> values are used when filtering the results. The q-value provides a measurement on the probability that the separation between the two conditions is due to chance, while the LOR measures the amount of separation. As a rule of thumb, we suggest considering as modified sites with <code>q-value &lt;= 0.01</code> <strong>and</strong> <code>|GMM_LOR| &gt;= 0.5</code> (i.e. absolute value of the LOR is larger than 0.5).</p>
<h2 id="shift-statistics-tsv">Shift statistics TSV</h2>
<p>The shift statistics TSV gives summary statistics (mean, median, standard deviation) at the position level for the signal measurements (current intensity and dwell time) for the two conditions. The data will always be gathered during the analysis and saved to the database, but it will be exported to a TSV file only when <code>export_shift_stats: True</code> is added to the configuration. The TSV will includes the following columns:</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>ref_id</strong></td>
<td>Reference name of the transcript.</td>
<td>ENST00000676788.1</td>
</tr>
<tr>
<td><strong>pos</strong></td>
<td>Starting position on the transcript's reference sequence of the k-mer for which the row reports results. Indexing is 0-based, meaning that the first nucleotide on the transcript is position 0.</td>
<td>42</td>
</tr>
<tr>
<td><strong>c1_mean_intensity</strong></td>
<td>Mean value for the current intensity at the position for condition 1.</td>
<td>78.91</td>
</tr>
<tr>
<td><strong>c2_mean_intensity</strong></td>
<td>Mean value for the current intensity at the position for condition 2.</td>
<td>81.91</td>
</tr>
<tr>
<td><strong>c1_median_intensity</strong></td>
<td>Median value for the current intensity at the position for condition 1.</td>
<td>78.21</td>
</tr>
<tr>
<td><strong>c2_median_intensity</strong></td>
<td>Median value for the current intensity at the position for condition 2.</td>
<td>80.21</td>
</tr>
<tr>
<td><strong>c1_sd_intensity</strong></td>
<td>Standard deviation for the current intensity at the position for condition 1.</td>
<td>1.21</td>
</tr>
<tr>
<td><strong>c2_sd_intensity</strong></td>
<td>Standard deviation for the current intensity at the position for condition 2.</td>
<td>2.21</td>
</tr>
<tr>
<td><strong>c1_mean_dwell</strong></td>
<td>Mean value for the dwell time at the position for condition 1.</td>
<td>0.31</td>
</tr>
<tr>
<td><strong>c2_mean_dwell</strong></td>
<td>Mean value for the dwell time at the position for condition 2.</td>
<td>0.23</td>
</tr>
<tr>
<td><strong>c1_median_dwell</strong></td>
<td>Median value for the dwell time at the position for condition 1.</td>
<td>0.29</td>
</tr>
<tr>
<td><strong>c2_median_dwell</strong></td>
<td>Median value for the dwell time at the position for condition 2.</td>
<td>0.33</td>
</tr>
<tr>
<td><strong>c1_sd_dwell</strong></td>
<td>Standard deviation for the dwell time at the position for condition 1.</td>
<td>0.19</td>
</tr>
<tr>
<td><strong>c2_sd_dwell</strong></td>
<td>Standard deviation for the dwell time at the position for condition 2.</td>
<td>0.23</td>
</tr>
</tbody>
</table>
<h2 id="result-database">Result database</h2>
<p>The TSV files described above are created for the user's convenience at the end of Nanocompore's run. All data for them are sourced from the SQLite database that Nanocompore uses throughout the run to store all results. The database would be found in the output directory under the filename "out_sampComp_sql.db".</p>
<p>The database schema is as follows:</p>
<div class="highlight"><pre><span></span><code><span class="k">CREATE</span><span class="w"> </span><span class="k">TABLE</span><span class="w"> </span><span class="n">transcripts</span><span class="w"> </span><span class="p">(</span>
<span class="w">    </span><span class="n">id</span><span class="w"> </span><span class="nb">INTEGER</span><span class="w"> </span><span class="k">PRIMARY</span><span class="w"> </span><span class="k">KEY</span><span class="w"> </span><span class="n">AUTOINCREMENT</span><span class="p">,</span>
<span class="w">    </span><span class="n">name</span><span class="w"> </span><span class="nb">VARCHAR</span><span class="w"> </span><span class="k">NOT</span><span class="w"> </span><span class="k">NULL</span><span class="w"> </span><span class="k">UNIQUE</span>
<span class="p">);</span>
<span class="k">CREATE</span><span class="w"> </span><span class="k">INDEX</span><span class="w"> </span><span class="n">transcripts_name_index</span>
<span class="w">    </span><span class="k">ON</span><span class="w"> </span><span class="n">transcripts</span><span class="p">(</span><span class="n">name</span><span class="p">);</span>

<span class="k">CREATE</span><span class="w"> </span><span class="k">TABLE</span><span class="w"> </span><span class="n">kmer_results</span><span class="w"> </span><span class="p">(</span>
<span class="w">    </span><span class="n">id</span><span class="w"> </span><span class="nb">INTEGER</span><span class="w"> </span><span class="k">PRIMARY</span><span class="w"> </span><span class="k">KEY</span><span class="w"> </span><span class="n">AUTOINCREMENT</span><span class="p">,</span>
<span class="w">    </span><span class="n">transcript_id</span><span class="w"> </span><span class="nb">INTEGER</span><span class="w"> </span><span class="k">NOT</span><span class="w"> </span><span class="k">NULL</span><span class="p">,</span>
<span class="w">    </span><span class="n">pos</span><span class="w"> </span><span class="nb">INTEGER</span><span class="w"> </span><span class="k">NOT</span><span class="w"> </span><span class="k">NULL</span><span class="p">,</span>
<span class="w">    </span><span class="n">kmer</span><span class="w"> </span><span class="nb">INTEGER</span><span class="w"> </span><span class="k">NOT</span><span class="w"> </span><span class="k">NULL</span><span class="p">,</span>
<span class="w">    </span><span class="n">c1_mean_intensity</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c1_mean_dwell</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c1_median_intensity</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c1_median_dwell</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c1_std_intensity</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c1_std_dwell</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c2_mean_intensity</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c2_mean_dwell</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c2_median_intensity</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c2_median_dwell</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c2_std_intensity</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">c2_std_dwell</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="k">UNIQUE</span><span class="w"> </span><span class="p">(</span><span class="n">transcript_id</span><span class="p">,</span><span class="w"> </span><span class="n">pos</span><span class="p">),</span>
<span class="w">    </span><span class="k">FOREIGN</span><span class="w"> </span><span class="k">KEY</span><span class="w"> </span><span class="p">(</span><span class="n">transcript_id</span><span class="p">)</span><span class="w"> </span><span class="k">REFERENCES</span><span class="w"> </span><span class="n">transcripts</span><span class="p">(</span><span class="n">id</span><span class="p">)</span>
<span class="p">);</span>
<span class="k">CREATE</span><span class="w"> </span><span class="k">INDEX</span><span class="w"> </span><span class="n">kmer_results_transcript_id_index</span>
<span class="w">    </span><span class="k">ON</span><span class="w"> </span><span class="n">kmer_results</span><span class="p">(</span><span class="n">transcript_id</span><span class="p">);</span>
</code></pre></div>
<p>However, depending on the choice of statistical tests, the samples used and other parameters, additional column may be added. For example, suppose we're using the GMM and KS tests for comparing 3 knock-down and 3 wilde type samples. We'd get the following additional columns in the <code>kmer_results</code> table.</p>
<div class="highlight"><pre><span></span><code><span class="k">CREATE</span><span class="w"> </span><span class="k">TABLE</span><span class="w"> </span><span class="n">kmer_results</span><span class="w"> </span><span class="p">(</span>
<span class="w">    </span><span class="p">...</span>
<span class="w">    </span><span class="c1">-- GMM columns</span>
<span class="w">    </span><span class="n">GMM_chi2_pvalue</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">GMM_LOR</span><span class="w"> </span><span class="nb">VARCHAR</span><span class="p">,</span>
<span class="w">    </span><span class="n">KD_1_mod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">KD_1_unmod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">KD_2_mod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">KD_2_unmod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">KD_3_mod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">KD_3_unmod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">WT_1_mod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">WT_1_unmod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">WT_2_mod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">WT_2_unmod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">WT_3_mod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">WT_3_unmod</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="c1">-- KS columns</span>
<span class="w">    </span><span class="n">KS_intensity_pvalue</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="n">KS_dwell_pvalue</span><span class="w"> </span><span class="nb">FLOAT</span><span class="p">,</span>
<span class="w">    </span><span class="p">...</span>
<span class="p">)</span>
</code></pre></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2019 Adrien Leger & Tommaso Leonardi
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>