
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Compare nanopore signal to find modifications">
      
      
        <meta name="author" content="Adrien Leger & Tommaso Leonardi">
      
      
        <link rel="canonical" href="http://nanocompore.rna.rocks/data_preparation/">
      
      
        <link rel="prev" href="../installation/">
      
      
        <link rel="next" href="../usage/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Data preparation - Nanocompore</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
      


    
    
      
    
    
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#data-preparation" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Nanocompore" class="md-header__button md-logo" aria-label="Nanocompore" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Nanocompore
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Data preparation
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/tleonardi/nanocompore" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    tleonardi/nanocompore
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Nanocompore" class="md-nav__button md-logo" aria-label="Nanocompore" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Nanocompore
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/tleonardi/nanocompore" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    tleonardi/nanocompore
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation instructions
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Data preparation
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Data preparation
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#reads-basecalling" class="md-nav__link">
    <span class="md-ellipsis">
      Reads basecalling
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#transcriptome-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Transcriptome alignment
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Transcriptome alignment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example-with-minimap2-v228-for-f5cs-eventalign" class="md-nav__link">
    <span class="md-ellipsis">
      Example with Minimap2 v2.28 for f5c's eventalign
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-with-minimap2-v228-for-uncalled4s-align" class="md-nav__link">
    <span class="md-ellipsis">
      Example with Minimap2 v2.28 for Uncalled4's align
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-with-minimap2-v228-for-remora" class="md-nav__link">
    <span class="md-ellipsis">
      Example with Minimap2 v2.28 for Remora
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resquiggling" class="md-nav__link">
    <span class="md-ellipsis">
      Resquiggling
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Resquiggling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#f5c" class="md-nav__link">
    <span class="md-ellipsis">
      f5c
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#uncalled" class="md-nav__link">
    <span class="md-ellipsis">
      Uncalled
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#remora" class="md-nav__link">
    <span class="md-ellipsis">
      Remora
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running Nanocompore
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Outputs
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Plotting
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data_api/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data API
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting_api/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Plotting API
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../troubleshooting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Troubleshooting
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../alternative/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Alternative and complementary packages
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../citing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../code_of_conduct/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Code of conduct
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../licence/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Licence
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../changelog/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Changelog
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#reads-basecalling" class="md-nav__link">
    <span class="md-ellipsis">
      Reads basecalling
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#transcriptome-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Transcriptome alignment
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Transcriptome alignment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example-with-minimap2-v228-for-f5cs-eventalign" class="md-nav__link">
    <span class="md-ellipsis">
      Example with Minimap2 v2.28 for f5c's eventalign
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-with-minimap2-v228-for-uncalled4s-align" class="md-nav__link">
    <span class="md-ellipsis">
      Example with Minimap2 v2.28 for Uncalled4's align
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-with-minimap2-v228-for-remora" class="md-nav__link">
    <span class="md-ellipsis">
      Example with Minimap2 v2.28 for Remora
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resquiggling" class="md-nav__link">
    <span class="md-ellipsis">
      Resquiggling
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Resquiggling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#f5c" class="md-nav__link">
    <span class="md-ellipsis">
      f5c
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#uncalled" class="md-nav__link">
    <span class="md-ellipsis">
      Uncalled
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#remora" class="md-nav__link">
    <span class="md-ellipsis">
      Remora
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="data-preparation">Data preparation</h1>
<p>Before using Nanocompore, sequencing data have to be basecalled (Dorado), aligned to a transcriptome reference (Minimap2), and resquiggled. Nanocompore supports multiple different resquigglers: Nanopolish/f5c eventalign, Uncalled4 align or Remora's API.</p>
<p>We're planning to release a Nextflow pipeline that would automate all those steps. However, the manual steps for performing the preprocessing are described below. The guide is based on <a href="https://doi.org/10.1002/cpz1.683">https://doi.org/10.1002/cpz1.683</a>. Note that this procedure should be repeated for all samples.</p>
<h3 id="reads-basecalling">Reads basecalling</h3>
<p>Firstly, raw pod5 reads have to be basecalled with a recent version of ONT basecaller.</p>
<p>Example with Dorado v0.8.0:</p>
<div class="highlight"><pre><span></span><code>dorado<span class="w"> </span>basecaller<span class="w"> </span>--emit-moves<span class="w"> </span><span class="o">{</span>dorado_model<span class="o">}</span><span class="w"> </span><span class="o">{</span>pod5<span class="o">}</span><span class="w"> </span>&gt;<span class="w"> </span><span class="o">{</span>basecalled_bam<span class="o">}</span>
</code></pre></div>
<h3 id="transcriptome-alignment">Transcriptome alignment</h3>
<p>Basecalled reads have to be aligned to a reference. For dRNA-Seq, reads should be aligned to a reference <strong>transcriptome (not genome)</strong> in a non-spliced fashion. For example, one can download reference transcriptome fasta files directly from Gencode for <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_47/gencode.v47.transcripts.fa.gz">human</a> and <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M36/gencode.vM36.transcripts.fa.gz">mouse</a>.</p>
<p>Bam files have to be filtered to remove any reads that would be unmapped, secondary and supplementary as well as reads mapped on the reverse strand (SAM flag 2324). We also recommend to discard reads with a very low alignment score (MAPQ&lt;10). Finally, reads have then to be sorted and indexed.</p>
<p>There are subtle differences in the process depending on the choice of resquiggler. Specifically, Uncalled4 requires extracting additional tags from the bam when creating a fasta file with the reads and then adding the <code>-y</code> option to Minimap2 to pass them to the aligned bam. Remora additonally requires adding the <code>--MD</code> flag to Minimap2. Examples for each resquiggler are given below.</p>
<h4 id="example-with-minimap2-v228-for-f5cs-eventalign">Example with <a href="https://github.com/lh3/minimap2">Minimap2 v2.28</a> for f5c's eventalign</h4>
<p>Extract a fasta file from the bam:</p>
<div class="highlight"><pre><span></span><code>samtools<span class="w"> </span>fasta<span class="w"> </span><span class="o">{</span>basecalled_bam<span class="o">}</span><span class="w"> </span>&gt;<span class="w"> </span><span class="o">{</span>reads_fasta<span class="o">}</span>

minimap2<span class="w"> </span>-ax<span class="w"> </span>map-ont<span class="w"> </span>-L<span class="w"> </span><span class="o">{</span>transcriptome_fasta<span class="o">}</span><span class="w"> </span><span class="o">{</span>reads_fasta<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>view<span class="w"> </span>-bh<span class="w"> </span>-F<span class="w"> </span><span class="m">2324</span><span class="w"> </span>-q<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>sort<span class="w"> </span>-O<span class="w"> </span>bam<span class="w"> </span>&gt;<span class="w"> </span><span class="o">{</span>aligned_reads_bam<span class="o">}</span>

samtools<span class="w"> </span>index<span class="w"> </span><span class="o">{</span>aligned_reads_bam<span class="o">}</span>
</code></pre></div>
<h4 id="example-with-minimap2-v228-for-uncalled4s-align">Example with <a href="https://github.com/lh3/minimap2">Minimap2 v2.28</a> for Uncalled4's align</h4>
<p>Extract a fasta file from the bam, adding the tags required by Uncalled4:</p>
<div class="highlight"><pre><span></span><code>samtools<span class="w"> </span>fasta<span class="w"> </span>-T<span class="w"> </span><span class="s2">&quot;mv,ts,pi,sp,ns&quot;</span><span class="w"> </span><span class="o">{</span>basecalled_bam<span class="o">}</span><span class="w"> </span>&gt;<span class="w"> </span><span class="o">{</span>reads_fasta<span class="o">}</span>

<span class="c1"># Note, we&#39;re adding the -y flag to pass the tags to the output bam</span>
minimap2<span class="w"> </span>-y<span class="w"> </span>-ax<span class="w"> </span>map-ont<span class="w"> </span>-L<span class="w"> </span><span class="o">{</span>transcriptome_fasta<span class="o">}</span><span class="w"> </span><span class="o">{</span>reads_fasta<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>view<span class="w"> </span>-bh<span class="w"> </span>-F<span class="w"> </span><span class="m">2324</span><span class="w"> </span>-q<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>sort<span class="w"> </span>-O<span class="w"> </span>bam<span class="w"> </span>&gt;<span class="w"> </span><span class="o">{</span>aligned_reads_bam<span class="o">}</span>

samtools<span class="w"> </span>index<span class="w"> </span><span class="o">{</span>aligned_reads_bam<span class="o">}</span>
</code></pre></div>
<h4 id="example-with-minimap2-v228-for-remora">Example with <a href="https://github.com/lh3/minimap2">Minimap2 v2.28</a> for Remora</h4>
<div class="highlight"><pre><span></span><code>samtools<span class="w"> </span>fasta<span class="w"> </span>-T<span class="w"> </span><span class="s2">&quot;mv,ts,pi,sp,ns&quot;</span><span class="w"> </span><span class="o">{</span>basecalled_bam<span class="o">}</span><span class="w"> </span>&gt;<span class="w"> </span><span class="o">{</span>reads_fasta<span class="o">}</span>

<span class="c1"># Note, we&#39;re adding the -y and --MD tags</span>
minimap2<span class="w"> </span>-y<span class="w"> </span>-ax<span class="w"> </span>map-ont<span class="w"> </span>-L<span class="w"> </span><span class="o">{</span>transcriptome_fasta<span class="o">}</span><span class="w"> </span><span class="o">{</span>reads_fasta<span class="o">}</span><span class="w"> </span>--MD<span class="w"> </span><span class="se">\</span>
<span class="w">  </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>view<span class="w"> </span>-bh<span class="w"> </span>-F<span class="w"> </span><span class="m">2324</span><span class="w"> </span>-q<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>sort<span class="w"> </span>-O<span class="w"> </span>bam<span class="w"> </span>&gt;<span class="w"> </span><span class="o">{</span>aligned_reads_bam<span class="o">}</span>

samtools<span class="w"> </span>index<span class="w"> </span><span class="o">{</span>aligned_reads_bam<span class="o">}</span>
</code></pre></div>
<h3 id="resquiggling">Resquiggling</h3>
<p>Nanocompore supports using signal alignments from various sources. Currently supported ones are:
- Eventalin TSVs as produced by <a href="https://github.com/jts/nanopolish">Nanopolish</a> or <a href="https://github.com/hasindu2008/f5c">f5c</a>. Since f5c provides optimized implementations of the eventalign command from Nanopolish, most users would be advised to use f5c.
- BAM files produced by <a href="https://github.com/skovaka/uncalled4">Uncalled4</a>.
- <a href="https://github.com/nanoporetech/remora">Remora</a> is a toolkit by ONT for training models for modification prediction. However, it also provides functionality for signal analysis, including capabilities to refine the signal alignments from the "moves" table produced by Dorado.</p>
<h4 id="f5c">f5c</h4>
<p>Since f5c doesn't support pod5, but uses the community-driven <code>slow5/blow5</code> formats, first you'll need to convert the pod5 to blow5 using <a href="https://github.com/Psy-Fer/blue-crab">blue-crab</a>.</p>
<p>Make sure that you have blue-crab installed. Typically, this should involve running:
<div class="highlight"><pre><span></span><code>pip<span class="w"> </span>install<span class="w"> </span>blue-crab
</code></pre></div></p>
<p>Convert the pod5 to blow5:
<div class="highlight"><pre><span></span><code>blue-crab<span class="w"> </span>p2s<span class="w"> </span>-o<span class="w"> </span><span class="o">{</span>blow5<span class="o">}</span><span class="w"> </span><span class="o">{</span>pod5<span class="o">}</span>
</code></pre></div>
Next, index the reads:
<div class="highlight"><pre><span></span><code>f5c<span class="w"> </span>index<span class="w"> </span>--slow5<span class="w"> </span><span class="o">{</span>blow5<span class="o">}</span><span class="w"> </span><span class="o">{</span>reads_fasta<span class="o">}</span>
</code></pre></div></p>
<p>Finally, we can resquiggle the reads. Nanocompore doesn't use the raw TSV file produced by the eventalign command for the comparison of the two conditions directly, but first preprocesses the TSV to obtain a collapsed version (with one row per k-mer). Since the raw TSV files are huge, we recommend that you directly pipe the output of f5c's eventalign to Nanocompore's eventalign_collapse subocommand.
<div class="highlight"><pre><span></span><code><span class="c1"># For RNA004 data</span>
f5c<span class="w"> </span>eventalign<span class="w"> </span>--bam<span class="w"> </span><span class="o">{</span>aligned_reads_bam<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--genome<span class="w"> </span><span class="o">{</span>transcriptome_fasta<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--reads<span class="w"> </span><span class="o">{</span>reads_fasta<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--slow5<span class="w"> </span><span class="o">{</span>blow5<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--scale-events<span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--print-read-name<span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--secondary<span class="o">=</span>yes<span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--min-mapq<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--samples<span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--summary<span class="w"> </span><span class="o">{</span>summary_file<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--rna<span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--pore<span class="w"> </span>rna004<span class="w"> </span><span class="se">\</span>
<span class="w">               </span>--threads<span class="w"> </span><span class="o">{</span>num_threads<span class="o">}</span>
<span class="w">  </span><span class="p">|</span><span class="w"> </span>nanocompore<span class="w"> </span>eventalign_collapse<span class="w"> </span>--ref<span class="w"> </span><span class="o">{</span>transcriptome_fasta<span class="o">}</span><span class="w"> </span>-o<span class="w"> </span><span class="o">{</span>resquiggled_db<span class="o">}</span><span class="w"> </span>--threads<span class="w"> </span><span class="o">{</span>num_threads<span class="o">}</span>
</code></pre></div>
Note that the eventalign_collapse creates rather big temporary files. By default it would use the current working directory. If you want to change that, you can add <code>--tmp path/to/tmp_dir</code>.</p>
<h4 id="uncalled">Uncalled</h4>
<p><div class="highlight"><pre><span></span><code><span class="c1"># For RNA004</span>
uncalled4<span class="w"> </span>align<span class="w"> </span>--ref<span class="w"> </span><span class="o">{</span>transcriptome_fasta<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">                </span>--reads<span class="w"> </span><span class="o">{</span>pod5<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">                </span>--bam-in<span class="w"> </span><span class="o">{</span>aligned_reads_bam<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">                </span>--bam-out<span class="w"> </span><span class="o">{</span>resquiggled_bam<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">                </span>--kit<span class="w"> </span>SQK-RNA004<span class="w"> </span><span class="se">\</span>
<span class="w">                </span>-p<span class="w"> </span><span class="o">{</span>num_threads<span class="o">}</span>
</code></pre></div>
This will produce a bam file with the signal alignment data that you can directly pass to Nanocompore for the analysis.</p>
<h4 id="remora">Remora</h4>
<p><a href="https://github.com/nanoporetech/remora">Remora</a> provides a Python API that can be used for resquiggling. Nanocompore provides a subcommand that will invoke Remora for aligning the signal. The command will produce an SQLite database with the signal data that you can then provide to Nanocompore for the analysis.</p>
<div class="highlight"><pre><span></span><code>nanocompore<span class="w"> </span>remora_resquiggle<span class="w"> </span>--ref<span class="w"> </span><span class="o">{</span>transcriptome_fasta<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">                              </span>--pod5<span class="w"> </span><span class="o">{</span>pod5<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">                              </span>--bam<span class="w"> </span><span class="o">{</span>aligned_reads_bam<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">                              </span>--kit<span class="w"> </span>RNA004<span class="w"> </span><span class="se">\</span>
<span class="w">                              </span>--output<span class="w"> </span><span class="o">{</span>resquiggled_db<span class="o">}</span>
</code></pre></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2019 Adrien Leger & Tommaso Leonardi
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>