
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Compare nanopore signal to find modifications">
      
      
        <meta name="author" content="Adrien Leger & Tommaso Leonardi">
      
      
        <link rel="canonical" href="http://nanocompore.rna.rocks/usage/">
      
      
        <link rel="prev" href="../data_preparation/">
      
      
        <link rel="next" href="../output/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Running Nanocompore - Nanocompore</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
      


    
    
      
    
    
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#how-to-use-nanocompore" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Nanocompore" class="md-header__button md-logo" aria-label="Nanocompore" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Nanocompore
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Running Nanocompore
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/tleonardi/nanocompore" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    tleonardi/nanocompore
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Nanocompore" class="md-nav__button md-logo" aria-label="Nanocompore" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Nanocompore
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/tleonardi/nanocompore" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    tleonardi/nanocompore
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation instructions
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../data_preparation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data preparation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Running Nanocompore
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Running Nanocompore
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#creating-a-configuration-file" class="md-nav__link">
    <span class="md-ellipsis">
      Creating a configuration file
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Creating a configuration file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#output-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Output configuration
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output configuration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#map-the-results-to-genomic-coordinates" class="md-nav__link">
    <span class="md-ellipsis">
      Map the results to genomic coordinates
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#shift-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      Shift statistics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#performance-tuning" class="md-nav__link">
    <span class="md-ellipsis">
      Performance tuning
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#continue-a-previous-run" class="md-nav__link">
    <span class="md-ellipsis">
      Continue a previous run
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Other parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-the-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Run the analysis
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Outputs
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Plotting
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data_api/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data API
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting_api/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Plotting API
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../troubleshooting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Troubleshooting
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../alternative/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Alternative and complementary packages
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../citing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../code_of_conduct/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Code of conduct
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../licence/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Licence
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../changelog/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Changelog
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#creating-a-configuration-file" class="md-nav__link">
    <span class="md-ellipsis">
      Creating a configuration file
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Creating a configuration file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#output-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Output configuration
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output configuration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#map-the-results-to-genomic-coordinates" class="md-nav__link">
    <span class="md-ellipsis">
      Map the results to genomic coordinates
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#shift-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      Shift statistics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#performance-tuning" class="md-nav__link">
    <span class="md-ellipsis">
      Performance tuning
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#continue-a-previous-run" class="md-nav__link">
    <span class="md-ellipsis">
      Continue a previous run
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Other parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-the-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Run the analysis
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="how-to-use-nanocompore">How to use Nanocompore</h1>
<p>Once all samples have been preprocessed as explained in the <a href="e/data_preparation">Data preparation</a> page, we can proceed to perform the comparison of the two conditions with Nanocompore.</p>
<h3 id="creating-a-configuration-file">Creating a configuration file</h3>
<p>Nanocompore uses a YAML configuration file that contains all parameters for a given analysis. We can generate a new configuration using the <strong>template</strong> subcommand:</p>
<div class="highlight"><pre><span></span><code>nanocompore<span class="w"> </span>template<span class="w"> </span>analysis.yaml
</code></pre></div>
<p>This will create a file called "analysis.yaml" in the current working directory. The file lists all parameters for Nanocompore documented with comments explaining their function and possible values. Here's the top part of the configuration file which describes the required fields:</p>
<p><div class="highlight"><pre><span></span><code><span class="c1"># === REQUIRED FIELDS ===</span>

<span class="c1"># Specify all the conditions and samples to bo analysed.</span>
<span class="c1"># Depending on the resquiggler used, Nanocompore expect different</span>
<span class="c1"># fields to be defined for all samples:</span>
<span class="c1"># - Uncalled4:</span>
<span class="c1">#   - Each sample should have the &quot;bam&quot; field set with a path</span>
<span class="c1">#     to a bam file containing the Uncalled4 resquiggling data.</span>
<span class="c1"># - Eventalign or Remora:</span>
<span class="c1">#   - Each sample should have the &quot;db&quot; field set with a path</span>
<span class="c1">#     to an SQLite database file produced by the subcommands</span>
<span class="c1">#     &quot;eventalign_collapse&quot; or &quot;remora_resquiggle&quot;.</span>
<span class="nt">data</span><span class="p">:</span>
<span class="w">  </span><span class="c1"># Only two conditions are supported. There&#39;s</span>
<span class="w">  </span><span class="c1"># no limit to the number of samples per condition.</span>
<span class="w">  </span><span class="c1"># You can change the labels appropriately, but</span>
<span class="w">  </span><span class="c1"># note that all labels should be unique.</span>
<span class="w">  </span><span class="nt">condition1</span><span class="p">:</span>
<span class="w">    </span><span class="nt">sample1</span><span class="p">:</span>
<span class="w">      </span><span class="nt">bam</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/path/to/uncalled4.bam&quot;</span>
<span class="w">      </span><span class="nt">db</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/path/to/collapsed_eventalign.sqlite&quot;</span>
<span class="w">    </span><span class="nt">sample2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">bam</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/path/to/uncalled4.bam&quot;</span>
<span class="w">      </span><span class="nt">db</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/path/to/collapsed_eventalign.sqlite&quot;</span>
<span class="w">    </span><span class="c1"># you can list as many samples as you want</span>
<span class="w">  </span><span class="nt">condition2</span><span class="p">:</span>
<span class="w">    </span><span class="nt">sample3</span><span class="p">:</span>
<span class="w">      </span><span class="nt">bam</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/path/to/uncalled4.bam&quot;</span>
<span class="w">      </span><span class="nt">db</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/path/to/collapsed_eventalign.sqlite&quot;</span>
<span class="w">    </span><span class="nt">sample4</span><span class="p">:</span>
<span class="w">      </span><span class="nt">bam</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/path/to/uncalled4.bam&quot;</span>
<span class="w">      </span><span class="nt">db</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/path/to/collapsed_eventalign.sqlite&quot;</span>

<span class="c1"># Which of the two conditions is the one that&#39;s</span>
<span class="c1"># depleted of modifications.</span>
<span class="nt">depleted_condition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">condition1</span>

<span class="c1"># Reference genome/transcriptome in fasta format</span>
<span class="nt">fasta</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/path/to/reference.fa&quot;</span>

<span class="c1"># The kit that was used for sequencing.</span>
<span class="c1"># Possible values: RNA002, RNA004</span>
<span class="nt">kit</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;RNA004&quot;</span>

<span class="c1"># Specify which resquiggler you have used for your data.</span>
<span class="c1"># Possible options are &quot;eventalign&quot;, &quot;uncalled4&quot;, and &quot;remora&quot;.</span>
<span class="nt">resquiggler</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;eventalign&#39;</span>
</code></pre></div>
The <code>data</code> field should list all samples, preprocessed with the same resquiggler. The <code>fasta</code> parameter should be set to the transcriptome reference (a FASTA file) that was used for the alignment of the reads.</p>
<p>Below, the file will list various optional parameters that the user can tweak. The most important ones are discussed below.</p>
<h4 id="output-configuration">Output configuration</h4>
<p>Nanocompore stores all results obtaned during the analysis in an SQLite database. After the analysis is done it will perform a postprocessing step in which it will do multiple-test correction (by default using the Benjamini-Hochberg method) and export the test results to an easy to read TSV file. One can request additional information to be exported using the configuration. The most important parameters are:</p>
<h5 id="map-the-results-to-genomic-coordinates">Map the results to genomic coordinates</h5>
<p>The results produced by Nanocompore will contain only transcriptomic coordinates by default (i.e. the transcript reference name and the position along the transcript). Nanocompore can optionally perform the mapping to genomic coordinates. To do this, simply add the following to the configuration:</p>
<div class="highlight"><pre><span></span><code><span class="nt">gtf</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;path/to/annotation.gtf&#39;</span>
</code></pre></div>
<p>Where the annotation should be compatible with the reference FASTA that is used, i.e. it should provide information for all references in the FASTA.</p>
<p>As a result, the columns <code>chr</code>, <code>strand</code>, and <code>genomicPos</code> in the results TSV file will be populated. Note that the <code>genomicPos</code> would use a 0-based indexing (the first base on the chromosome has index 0).</p>
<h5 id="shift-statistics">Shift statistics</h5>
<p>The shift statistics TSV gives summary statistics (mean, median, standard deviation) at the position level for the signal measurements (current intensity and dwell time) for the two conditions (see <a href="/output">Outputs</a>). To enable the exporting of this file just add the folling to the configuration:</p>
<div class="highlight"><pre><span></span><code><span class="nt">export_shift_stats</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
</code></pre></div>
<h4 id="performance-tuning">Performance tuning</h4>
<p>The main way to control the performance and resource usage of Nanocompore is through the <code>devices</code> parameter.</p>
<p><div class="highlight"><pre><span></span><code><span class="c1"># Control how many worker processes to run and which</span>
<span class="c1"># computing device they should run on.</span>
<span class="c1"># The parameter should be a map of devices and number of worker processes.</span>
<span class="c1"># The accepted device values are: &quot;cpu&quot;, &quot;cuda&quot;, &quot;cuda:N&quot;</span>
<span class="c1"># For example, to use 4 worker processes on the CPU and</span>
<span class="c1"># 8 worker processes on each of the two available GPUs</span>
<span class="c1"># one can set the following:</span>
<span class="nt">devices</span><span class="p">:</span>
<span class="w">  </span><span class="s">&quot;cpu&quot;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">4</span>
<span class="w">  </span><span class="s">&quot;cuda:0&quot;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">8</span>
<span class="w">  </span><span class="s">&quot;cuda:1&quot;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">8</span>
</code></pre></div>
The <code>devices</code> parameter specifies how many worker processes Nanocompore should run. Typically, the more parallelization we use, the faster the analysis will go, as long as this matches the computational resources available. However, increasing the number of workers means that more memory would be used, potentially leading to out-of-memory errors. Some guidelines to consider:</p>
<ul>
<li>Running the analysis on GPU devices is significantly faster than using CPUs.</li>
<li>If using CPUs, a reasonable starting point is setting the parameter to <code>"cpu": N-1</code> where N is the number of available CPU cores. This would allow each worker to utilize one core while leaving one core for the main orchestrator process.</li>
<li>If using GPUs, the user is advised to try experimenting with different number of workers to tune the performance according to their GPU model. In particular, one should increase the worker count to achieve high GPU utilization, while making sure that the GPU memory consumption is low enough to avoid out-of-memory errors. Each worker process would have a copy of the CUDA context, which takes a few hundred MB of memory. On top of that, the worker will load to the GPU memory the data for the transcript that it currently is processing. Monitoring the run logs can show you if out of memory errors occur. In the case of such error, Nanocompore will add the transcript back to the queue to retry it later (up to 3 times), so an occasional out-of-memory is not a reason to be very worried. However, if they are too frequent it means that the GPU is overloaded which would lead to lower performance and potential loss of results.</li>
<li>For example, for a GPU model NVIDIA Tesla P100-PCIE, which has 12GB of memory we typically use 8 workers per GPU device.</li>
<li>You can mix devices to improve the performance. However, bear in mind that GPU workers would also use the CPU to some degree. Hence, it's not recommended to use all CPU cores with CPU workers if you also run GPU workers. E.g. if you have 32 cores and you run 16 GPU workers, you can try running 15 CPU workers on top, but it's not wise to try running 32 CPU workers.</li>
</ul>
<h4 id="continue-a-previous-run">Continue a previous run</h4>
<p>By default, if the output directory for the run already exists, Nanocompore will stop and output an error message. You can instruct Nanocompore how to proceed in such cases by changing the parameter <code>result_exists_strategy</code>: </p>
<p><div class="highlight"><pre><span></span><code><span class="c1"># What to do if the output directory already exists.</span>
<span class="c1"># Possible options are:</span>
<span class="c1"># - stop (default): stop the execution with an error.</span>
<span class="c1"># - overwrite: delete the existing directory and start from scratch.</span>
<span class="c1"># - continue: try to continue from where the previous execution left off.</span>
<span class="nt">result_exists_strategy</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;continue&quot;</span>
</code></pre></div>
If you set the parameter to "continue", Nanocompore will try to reuse the database that was created in the previous run and won't repeat the analysis for any transcripts that have already been analysed and written to the database.</p>
<h4 id="other-parameters">Other parameters</h4>
<p>For a full list of parameters and a description of their function, inspect a template configuration created by the <code>nanocompore template</code> command.</p>
<h3 id="run-the-analysis">Run the analysis</h3>
<p>Once you have created a configuration, you can run the analysis on your data with:</p>
<div class="highlight"><pre><span></span><code>nanocompore<span class="w"> </span>run<span class="w"> </span>analysis.yaml
</code></pre></div>
<p>Nanocompore should immediately create a directory "nanocompore_output" in the current working directory (unless you have set the <code>outpath</code> parameter in the configuration file to a different location). All working files, logs, and results will be placed in that directory.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2019 Adrien Leger & Tommaso Leonardi
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>