{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Eventalign_collapse usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## In brief"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This program collapses the raw file generated by `nanopolish eventalign` by kmers rather than by event."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quick start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-07T19:06:20.606725Z",
     "start_time": "2020-12-07T19:06:20.590584Z"
    }
   },
   "source": [
    "Example CLI call\n",
    "\n",
    "```bash\n",
    "nanocompore eventalign_collapse \\\n",
    "    -i eventalign_file.tsv\n",
    "    --outpath results\n",
    "```\n",
    "\n",
    "Example API call\n",
    "\n",
    "\n",
    "\n",
    "```python\n",
    "# Import package\n",
    "from nanocompore.Eventalign_collapse import Eventalign_collapse\n",
    "\n",
    "# Init the object\n",
    "e = Eventalign_collapse(\n",
    "    eventalign_fn=\"eventalign_file.tsv\",\n",
    "    outpath=\"results\")\n",
    "\n",
    "# Run the analysis\n",
    "e()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Description of options"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Input nanopolish eventalign file(s)/stream\n",
    "\n",
    "Path to a `nanopolish eventalign` tsv output file (read access required) which can optionally be gzipped.\n",
    "This can also be a list of files, or a regex or a list of regex or a mix of everything. We are flexible.\n",
    "In command line mode it is also possible to pipe the output of the `nanopolish eventalign` directly into `Eventalign_collapse`.\n",
    "\n",
    "#### Output options\n",
    "\n",
    "`outdir` is a path to the directory where to write all the output files generated by `Eventalign_collapse`(write access required, created if it does not exist).\n",
    "If it already exists, an error will be raise except if `overwrite` is given. `outprefix` is the file prefix used for all files generated.\n",
    "  \n",
    "#### n_lines\n",
    "\n",
    "Controls the maximum number of lines to parse before stopping to read the input file. This could be useful for testing or downsampling. \n",
    "\n",
    "#### threads\n",
    "\n",
    "`Eventalign_collapse` is multi threaded to speed up the data processing and keep pace with Nanopolish if using the direct piping strategy. Take advantage of many threads if you have access to a large compute cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output files format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Contrary to `nanopolish eventalign` output text file, in `Eventalign_collapse` the reads are separated by a hashtag headers containing the read_id and ref_id. This reduces the redundancy and makes it easier to find the start and end of a read.\n",
    "\n",
    "Example : ```#7ef1d7b9-5824-4382-b23b-78d82c07ebbd\tYHR055C.```\n",
    "\n",
    "The main data file contains the following fields:\n",
    "\n",
    "ref_pos\tref_kmer\tnum_events\tdwell_time\tNNNNN_dwell_time\tmismatch_dwell_time\tstatus\tmedian\tmad\tnum_signals\n",
    "\n",
    "* **ref_pos**: Reference sequence ID (contig).\n",
    "* **ref_kmer**: Sequence of the reference kmers.\n",
    "* **num_events**: Number of events for this kmer before collapsing.\n",
    "* **num_signals**: Number of raw signal data points.\n",
    "* **dwell_time**: dwell time for this kmer in seconds\n",
    "* **NNNNN_dwell_time**: dwell time of events for this kmers with a model sequence \"NNNNN\" (events ignored by nanopolish HMM).\n",
    "* **mismatch_dwell_time**:  dwell time of events for this kmers with a model sequence different from the reference kmer\n",
    "* **status**: Status of the kmer. `valid` if all the model and predicted kmers are the same for all events. Otherwise `NNNNN` or `mismatching` depending on the invalid state having the most signal points. \n",
    "* **median**: Median of the normalised intensity values provided by Nanopolish eventalign\n",
    "* **mad**: Median absolute deviation of the normalised intensity values provided by Nanopolish eventalign\n",
    "\n",
    "In addition `Eventalign_collapse` also generates a useful index file containing reads level information. It contains the following fields:\n",
    "\n",
    "num_events num_signals kmers missing_kmers NNNNN_kmers mismatch_kmers valid_kmers byte_offset byte_len \n",
    "\n",
    "* **ref_id**: Name of the reference sequence the read was aligned on (contig)\n",
    "* **ref_start**: Start coordinate of the alignment on the reference sequence\n",
    "* **ref_end**: End coordinate of the alignment on the reference sequence\n",
    "* **read_id**: Name or index of the read\n",
    "* **num_events**: Total number of events per read \n",
    "* **num_signals**: Total number of signal points per read\n",
    "* **dwell_time**: Cumulative dwell time in seconds for the entire resquiggled sequence\n",
    "* **kmers**: Overall number of kmers in the read, including the missing ones\n",
    "* **NNNNN_kmers**: Number of resquiggled kmers for which the majority of missing data points correspond to \"NNNNN\" kmers\n",
    "* **mismatching_kmers**:Number of resquiggled kmers for which the majority of missing data points correspond to kmers different from the model kmer (excluding NNNNN)\n",
    "* **missing_kmers**: Number of skipped/missing reference positions in nanopolish output\n",
    "* **valid_kmers**: Number of fully valid kmers\n",
    "* **byte_offset**: Number of characters before the start of the sequence in the main output file. **This can be used in conjunction with file.seek() to directly access the start of a read**. An example is provided in the Usage notebook.\n",
    "* **byte_len**: Length of characters after byte_offset to the end of the read, excluding the last newline. **This can be used in conjunction with read() to read all the text chunk corresponding to the read**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import the package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-07T22:07:37.933577Z",
     "start_time": "2020-12-07T22:07:37.351143Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating /home/aleg/Programming/Packages/nanocompore/docs/demo/results\n"
     ]
    }
   ],
   "source": [
    "from nanocompore.Eventalign_collapse import Eventalign_collapse\n",
    "from nanocompore.common import *\n",
    "from pycltools.pycltools import *\n",
    "import shutil\n",
    "\n",
    "# Create output dir and define log level\n",
    "outdir = \"results\"\n",
    "shutil.rmtree (outdir)\n",
    "mkdir(outdir, exist_ok=True)\n",
    "set_logger(\"info\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example with minimal file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-07T22:07:39.912649Z",
     "start_time": "2020-12-07T22:07:39.516274Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-12-07T22:07:39.526605+0000 INFO - MainProcess | Checking and initialising Eventalign_collapse\n",
      "2020-12-07T22:07:39.528104+0000 INFO - MainProcess | Starting data processing\n",
      "28 reads [00:00, 94.18 reads/s]\n",
      "2020-12-07T22:07:39.900955+0000 INFO - Process-4 | Output reads written:28\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#c96c0f7b-9596-43c5-a545-8e15a8a7f368\tcontrol\n",
      "ref_pos\tref_kmer\tnum_events\tnum_signals\tdwell_time\tNNNNN_dwell_time\tmismatch_dwell_time\tstatus\tmedian\tmad\n",
      "1\tTACTC\t1\t22\t0.007300000172108412\t0.0\t0.0\tvalid\t96.916504\t0.8642044\n",
      "2\tACTCG\t1\t12\t0.003980000037699938\t0.0\t0.0\tvalid\t95.18815\t0.864151\n",
      "3\tCTCGA\t1\t25\t0.008299999870359898\t0.0\t0.0\tvalid\t85.6821\t1.5363007\n",
      "4\tTCGAC\t2\t61\t0.02026000013574958\t0.0\t0.0\tvalid\t96.4364\t1.9204025\n",
      "5\tCGACA\t5\t77\t0.025550000369548798\t0.0\t0.0\tvalid\t110.84\t4.2249985\n",
      "6\tGACAT\t1\t12\t0.003980000037699938\t0.0\t0.0\tvalid\t89.23485\t0.7681503\n",
      "7\tACATA\t1\t12\t0.003980000037699938\t0.0\t0.0\tvalid\t86.6423\t1.3443031\n",
      "8\tCATAG\t3\t52\t0.017259999876841903\t0.015269999857991934\t0.0\tNNNNN\t152.513\t6.1455\n",
      "\n",
      "ref_id  ref_start ref_end read_id                              num_events num_signals dwell_time         kmers missing_kmers NNNNN_kmers mismatch_kmers valid_kmers byte_offset byte_len \n",
      "control 1         85      c96c0f7b-9596-43c5-a545-8e15a8a7f368 414        7781        2.5830999956233427 85    34            8           0              43          0           3673     \n",
      "control 1         89      fd13d01a-95ef-4073-b9cf-fb7277da2c04 453        8578        2.8478099999483675 89    40            8           0              41          3674        3546     \n",
      "control 1         87      b5783126-4472-45b4-9414-322a4a1a1776 540        10909       3.6216799857793376 87    32            10          0              45          7221        4019     \n",
      "control 1         90      778ba618-ced5-4e9e-81f8-16419d209940 201        4132        1.3718199981376529 90    1             3           0              86          11241       6125     \n",
      "control 2         88      506ff40c-92c3-4ff5-9169-3fbb9de2fc6b 301        6408        2.127389994333498  87    45            3           0              39          17367       3014     \n",
      "control 2         80      0402647b-c92d-4b5d-a820-2ce437a4ec6f 171        3288        1.0915600005537271 79    2             1           0              76          20382       5263     \n",
      "control 1         88      7f5e9d3b-60a5-4a27-9d74-053cc7f4791a 540        10888       3.6146999960765243 88    32            10          0              46          25646       4058     \n",
      "control 2         87      97166ef2-5d04-4544-9b6d-f3084aff6951 326        6433        2.135599997243844  86    34            5           0              47          29705       3716     \n",
      "control 4         88      3127653f-163f-4fcc-bcb4-90ec8ea7e638 357        7978        2.6487000052584335 85    29            4           0              52          33422       3953     \n",
      "\n"
     ]
    }
   ],
   "source": [
    "e = Eventalign_collapse(\n",
    "    eventalign_fn=\"./nanopolish_eventalign/nanopolish_reads_2.tsv\",\n",
    "    outpath=\"results\",\n",
    "    outprefix=\"s2\",\n",
    "    overwrite=True,\n",
    "    progress=True,\n",
    "    nthreads=4)\n",
    "e()\n",
    "\n",
    "head(\"./results/s2_eventalign_collapse.tsv\")\n",
    "head(\"./results/s2_eventalign_collapse.tsv.idx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using the index to random access a specific entry in the file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Random access with standard library  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-07T22:08:58.188539Z",
     "start_time": "2020-12-07T22:08:58.135564Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RANDOM INDEX LINES\n",
      "ref_id      ref_start    ref_end  read_id                                 num_events    num_signals    dwell_time    kmers    missing_kmers    NNNNN_kmers    mismatch_kmers    valid_kmers    byte_offset    byte_len\n",
      "--------  -----------  ---------  ------------------------------------  ------------  -------------  ------------  -------  ---------------  -------------  ----------------  -------------  -------------  ----------\n",
      "control            10         89  6b53dc3c-4913-4970-bc92-98c8a28d510a           299           6562       2.17847       80               26              9                 0             45          81568        3895\n",
      "control             2         80  0402647b-c92d-4b5d-a820-2ce437a4ec6f           171           3288       1.09156       79                2              1                 0             76          20382        5263\n",
      "control             8         89  465fdd43-e5be-4c66-9165-3ba9e2817109           365           7285       2.4185        82               21              6                 0             55          67007        4332\n",
      "control             9         80  0c08b5d8-6c89-4ca6-b4dc-202096535762           178           3627       1.20414       72               13              6                 0             53          85464        4180\n",
      "control             1         85  c96c0f7b-9596-43c5-a545-8e15a8a7f368           414           7781       2.5831        85               34              8                 0             43              0        3673\n",
      "\n",
      "#6b53dc3c-4913-4970-bc92-98c8a28d510a\tcontrol\n",
      "  ref_pos  ref_kmer      num_events    num_signals    dwell_time    NNNNN_dwell_time    mismatch_dwell_time  status      median      mad\n",
      "---------  ----------  ------------  -------------  ------------  ------------------  ---------------------  --------  --------  -------\n",
      "       10  TAGAT                  1              8       0.00266             0                            0  valid     153.746   10.5975\n",
      "       11  AGATA                  4             57       0.01892             0                            0  valid     138.135   10.027\n",
      "       12  GATAG                  4             95       0.03155             0.00332                      0  NNNNN     107.141   15.725\n",
      "       14  TAGGA                 10            359       0.11919             0                            0  valid      97.3414   6.837\n",
      "       15  AGGAC                  2             40       0.01328             0                            0  valid     114.434    5.6975\n",
      "\n",
      "#0402647b-c92d-4b5d-a820-2ce437a4ec6f\tcontrol\n",
      "  ref_pos  ref_kmer      num_events    num_signals    dwell_time    NNNNN_dwell_time    mismatch_dwell_time  status      median     mad\n",
      "---------  ----------  ------------  -------------  ------------  ------------------  ---------------------  --------  --------  ------\n",
      "        2  ACTCG                  2             57       0.01892                   0                      0  valid      83.736   1.8462\n",
      "        3  CTCGA                  1             34       0.01129                   0                      0  valid      91.6244  2.1819\n",
      "        4  TCGAC                  2             48       0.01593                   0                      0  valid      93.3028  2.2658\n",
      "        5  CGACA                  1             72       0.0239                    0                      0  valid     108.073   6.6295\n",
      "        6  GACAT                  1             17       0.00564                   0                      0  valid      83.2325  1.6784\n",
      "\n",
      "#465fdd43-e5be-4c66-9165-3ba9e2817109\tcontrol\n",
      "  ref_pos  ref_kmer      num_events    num_signals    dwell_time    NNNNN_dwell_time    mismatch_dwell_time  status      median      mad\n",
      "---------  ----------  ------------  -------------  ------------  ------------------  ---------------------  --------  --------  -------\n",
      "        8  CATAG                  1              7       0.00232                   0                      0  valid      94.5734  1.5419\n",
      "        9  ATAGA                  1             10       0.00332                   0                      0  valid     118.087   3.2765\n",
      "       10  TAGAT                  1             33       0.01096                   0                      0  valid     138.903   2.50501\n",
      "       11  AGATA                  5            149       0.04947                   0                      0  valid     136.205   3.084\n",
      "       12  GATAG                  1             10       0.00332                   0                      0  valid     106.62    5.3005\n",
      "\n",
      "#0c08b5d8-6c89-4ca6-b4dc-202096535762\tcontrol\n",
      "  ref_pos  ref_kmer      num_events    num_signals    dwell_time    NNNNN_dwell_time    mismatch_dwell_time  status      median      mad\n",
      "---------  ----------  ------------  -------------  ------------  ------------------  ---------------------  --------  --------  -------\n",
      "        9  ATAGA                  2             17       0.00565             0                            0  valid      98.6508  9.0242\n",
      "       10  TAGAT                  3             51       0.01693             0.00232                      0  NNNNN     115.811   4.14201\n",
      "       14  TAGGA                 13            248       0.08232             0                            0  valid     114.183   6.287\n",
      "       15  AGGAC                  5            179       0.05943             0                            0  valid     116.107   4.734\n",
      "       16  GGACT                 14            358       0.11886             0                            0  valid     126.905   3.55\n",
      "\n",
      "#c96c0f7b-9596-43c5-a545-8e15a8a7f368\tcontrol\n",
      "  ref_pos  ref_kmer      num_events    num_signals    dwell_time    NNNNN_dwell_time    mismatch_dwell_time  status      median       mad\n",
      "---------  ----------  ------------  -------------  ------------  ------------------  ---------------------  --------  --------  --------\n",
      "        1  TACTC                  1             22       0.0073                    0                      0  valid      96.9165  0.864204\n",
      "        2  ACTCG                  1             12       0.00398                   0                      0  valid      95.1881  0.864151\n",
      "        3  CTCGA                  1             25       0.0083                    0                      0  valid      85.6821  1.5363\n",
      "        4  TCGAC                  2             61       0.02026                   0                      0  valid      96.4364  1.9204\n",
      "        5  CGACA                  5             77       0.02555                   0                      0  valid     110.84    4.225\n"
     ]
    }
   ],
   "source": [
    "output_fn = \"./results/s2_eventalign_collapse.tsv\"\n",
    "index_fn = output_fn+\".idx\"\n",
    "\n",
    "# Imports\n",
    "import csv\n",
    "from tabulate import tabulate\n",
    "from random import sample\n",
    "from itertools import islice\n",
    "\n",
    "# read index file and select random lines\n",
    "index_list = []\n",
    "with open (index_fn) as fp:\n",
    "    for l in csv.DictReader(fp, delimiter='\\t'):\n",
    "        index_list.append(l)\n",
    "random_index = sample(index_list, k=5)\n",
    "print (\"RANDOM INDEX LINES\")\n",
    "print (tabulate(random_index, headers=\"keys\"))\n",
    "\n",
    "# Open the collapsed event align file\n",
    "with open (output_fn) as fp:\n",
    "    for index in random_index:\n",
    "        # Access the header corresponding to the randomly selected index line using seek \n",
    "        fp.seek(0) # Return to file start\n",
    "        fp.seek(int(index[\"byte_offset\"])) # Move to the offset indicated in the index file\n",
    "        print (\"\\n\" + fp.readline().rstrip()) # Print read header\n",
    "        # Get 5 first lines and print then\n",
    "        data_list = []\n",
    "        for l in islice(csv.DictReader(fp, delimiter='\\t'), 5):\n",
    "            data_list.append(l)\n",
    "        print (tabulate(data_list, headers=\"keys\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example of random access with pandas  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-07T22:07:44.631452Z",
     "start_time": "2020-12-07T22:07:44.442189Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Random index lines\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_id</th>\n",
       "      <th>ref_start</th>\n",
       "      <th>ref_end</th>\n",
       "      <th>...</th>\n",
       "      <th>valid_kmers</th>\n",
       "      <th>byte_offset</th>\n",
       "      <th>byte_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>control</td>\n",
       "      <td>8</td>\n",
       "      <td>89</td>\n",
       "      <td>...</td>\n",
       "      <td>55</td>\n",
       "      <td>67007</td>\n",
       "      <td>4332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>control</td>\n",
       "      <td>4</td>\n",
       "      <td>92</td>\n",
       "      <td>...</td>\n",
       "      <td>51</td>\n",
       "      <td>37376</td>\n",
       "      <td>4138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>control</td>\n",
       "      <td>9</td>\n",
       "      <td>85</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>103487</td>\n",
       "      <td>3341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>control</td>\n",
       "      <td>9</td>\n",
       "      <td>86</td>\n",
       "      <td>...</td>\n",
       "      <td>41</td>\n",
       "      <td>89645</td>\n",
       "      <td>3670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>control</td>\n",
       "      <td>1</td>\n",
       "      <td>85</td>\n",
       "      <td>...</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>3673</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref_id  ref_start  ref_end  ... valid_kmers  byte_offset  byte_len\n",
       "16  control          8       89  ...          55        67007      4332\n",
       "9   control          4       92  ...          51        37376      4138\n",
       "26  control          9       85  ...          39       103487      3341\n",
       "22  control          9       86  ...          41        89645      3670\n",
       "0   control          1       85  ...          43            0      3673\n",
       "\n",
       "[5 rows x 14 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#465fdd43-e5be-4c66-9165-3ba9e2817109\tcontrol\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>...</th>\n",
       "      <th>status</th>\n",
       "      <th>median</th>\n",
       "      <th>mad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>CATAG</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>94.573400</td>\n",
       "      <td>1.541901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>ATAGA</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>118.087494</td>\n",
       "      <td>3.276501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>TAGAT</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>138.903000</td>\n",
       "      <td>2.505005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>AGATA</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>136.205000</td>\n",
       "      <td>3.084000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12</td>\n",
       "      <td>GATAG</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>106.619500</td>\n",
       "      <td>5.300503</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_pos ref_kmer  num_events  ...  status      median       mad\n",
       "0        8    CATAG           1  ...   valid   94.573400  1.541901\n",
       "1        9    ATAGA           1  ...   valid  118.087494  3.276501\n",
       "2       10    TAGAT           1  ...   valid  138.903000  2.505005\n",
       "3       11    AGATA           5  ...   valid  136.205000  3.084000\n",
       "4       12    GATAG           1  ...   valid  106.619500  5.300503\n",
       "\n",
       "[5 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#73f1c3ca-2bd3-4543-9b34-693935dce530\tcontrol\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>...</th>\n",
       "      <th>status</th>\n",
       "      <th>median</th>\n",
       "      <th>mad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>TCGAC</td>\n",
       "      <td>44</td>\n",
       "      <td>...</td>\n",
       "      <td>NNNNN</td>\n",
       "      <td>95.305800</td>\n",
       "      <td>7.780205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>CGACA</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>121.944500</td>\n",
       "      <td>11.669998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>GACAT</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>75.939900</td>\n",
       "      <td>0.930199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>ACATA</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>77.039300</td>\n",
       "      <td>2.452454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>CATAG</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>92.515045</td>\n",
       "      <td>0.845703</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_pos ref_kmer  num_events  ...  status      median        mad\n",
       "0        4    TCGAC          44  ...   NNNNN   95.305800   7.780205\n",
       "1        5    CGACA          15  ...   valid  121.944500  11.669998\n",
       "2        6    GACAT           1  ...   valid   75.939900   0.930199\n",
       "3        7    ACATA           4  ...   valid   77.039300   2.452454\n",
       "4        8    CATAG           1  ...   valid   92.515045   0.845703\n",
       "\n",
       "[5 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#2cc2d044-d8d9-4e72-a616-a0d25a787b2d\tcontrol\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>...</th>\n",
       "      <th>status</th>\n",
       "      <th>median</th>\n",
       "      <th>mad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>ATAGA</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NNNNN</td>\n",
       "      <td>82.00195</td>\n",
       "      <td>2.714901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>ATAGG</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>82.24150</td>\n",
       "      <td>2.714897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>TAGGA</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>109.31050</td>\n",
       "      <td>4.791000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>AGGAC</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>NNNNN</td>\n",
       "      <td>118.49300</td>\n",
       "      <td>5.828999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>NNNNN</td>\n",
       "      <td>126.79800</td>\n",
       "      <td>4.152000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_pos ref_kmer  num_events  ...  status     median       mad\n",
       "0        9    ATAGA           1  ...   NNNNN   82.00195  2.714901\n",
       "1       13    ATAGG           6  ...   valid   82.24150  2.714897\n",
       "2       14    TAGGA           1  ...   valid  109.31050  4.791000\n",
       "3       15    AGGAC          17  ...   NNNNN  118.49300  5.828999\n",
       "4       16    GGACT          45  ...   NNNNN  126.79800  4.152000\n",
       "\n",
       "[5 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#6f044625-cf92-4819-9e1d-31757d003355\tcontrol\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>...</th>\n",
       "      <th>status</th>\n",
       "      <th>median</th>\n",
       "      <th>mad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>ATAGA</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>109.451000</td>\n",
       "      <td>13.992195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>TAGAT</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>118.098495</td>\n",
       "      <td>6.315506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>AGATA</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>134.714000</td>\n",
       "      <td>6.024002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>GATAG</td>\n",
       "      <td>23</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>98.568100</td>\n",
       "      <td>1.360298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>ATAGG</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>84.964900</td>\n",
       "      <td>3.109299</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_pos ref_kmer  num_events  ...  status      median        mad\n",
       "0        9    ATAGA           4  ...   valid  109.451000  13.992195\n",
       "1       10    TAGAT           7  ...   valid  118.098495   6.315506\n",
       "2       11    AGATA           4  ...   valid  134.714000   6.024002\n",
       "3       12    GATAG          23  ...   valid   98.568100   1.360298\n",
       "4       13    ATAGG          11  ...   valid   84.964900   3.109299\n",
       "\n",
       "[5 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#c96c0f7b-9596-43c5-a545-8e15a8a7f368\tcontrol\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>...</th>\n",
       "      <th>status</th>\n",
       "      <th>median</th>\n",
       "      <th>mad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TACTC</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>96.916504</td>\n",
       "      <td>0.864204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>ACTCG</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>95.188150</td>\n",
       "      <td>0.864151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>CTCGA</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>85.682100</td>\n",
       "      <td>1.536301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>TCGAC</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>96.436400</td>\n",
       "      <td>1.920403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CGACA</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>valid</td>\n",
       "      <td>110.840000</td>\n",
       "      <td>4.224998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_pos ref_kmer  num_events  ...  status      median       mad\n",
       "0        1    TACTC           1  ...   valid   96.916504  0.864204\n",
       "1        2    ACTCG           1  ...   valid   95.188150  0.864151\n",
       "2        3    CTCGA           1  ...   valid   85.682100  1.536301\n",
       "3        4    TCGAC           2  ...   valid   96.436400  1.920403\n",
       "4        5    CGACA           5  ...   valid  110.840000  4.224998\n",
       "\n",
       "[5 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "output_fn = \"./results/s2_eventalign_collapse.tsv\"\n",
    "index_fn = output_fn+\".idx\"\n",
    "\n",
    "# Imports\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 6)\n",
    "\n",
    "# read index file and select random lines\n",
    "index_df = pd.read_csv (index_fn, sep=\"\\t\")\n",
    "random_lines = index_df.sample(5)\n",
    "print (\"Random index lines\")\n",
    "display (random_lines)\n",
    "\n",
    "# Open the collapsed event align file\n",
    "with open (output_fn) as fp:\n",
    "    for id, read in random_lines.iterrows():\n",
    "        \n",
    "        # Access the header corresponding to the randomly selected index line using seek \n",
    "        fp.seek(0) # Return to file start\n",
    "        fp.seek(read.byte_offset) # Move to the offset indicated in the index file\n",
    "        print (fp.readline().rstrip()) # Print read header\n",
    "        df = pd.read_csv (fp, nrows=5, sep=\"\\t\") # Read lines corresponding to the read\n",
    "        display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Full CLI and API documentations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### API documentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-08T11:26:55.528620Z",
     "start_time": "2020-12-08T11:26:55.517376Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**__init__** (eventalign_fn, outpath, outprefix, overwrite, n_lines, nthreads, progress)\n",
       "\n",
       "Collapse the nanopolish eventalign events at kmer level\n",
       "\n",
       "---\n",
       "\n",
       "* **eventalign_fn** (required) [str]\n",
       "\n",
       "Path to a nanopolish eventalign tsv output file, or a list of file, or a regex (can be gzipped)\n",
       "\n",
       "* **outpath** (default: ./) [str]\n",
       "\n",
       "Path to the output folder (will be created if it does exist yet)\n",
       "\n",
       "* **outprefix** (default: out) [str]\n",
       "\n",
       "text outprefix for all the files generated\n",
       "\n",
       "* **overwrite** (default: False) [bool]\n",
       "\n",
       "If the output directory already exists, the standard behaviour is to raise an error to prevent overwriting existing data This option ignore the error and overwrite data if they have the same outpath and outprefix.\n",
       "\n",
       "* **n_lines** (default: None) [int]\n",
       "\n",
       "Maximum number of read to parse.\n",
       "\n",
       "* **nthreads** (default: 3) [int]\n",
       "\n",
       "Number of threads (two are used for reading and writing, all the others for parallel processing).\n",
       "\n",
       "* **progress** (default: False) [bool]\n",
       "\n",
       "Display a progress bar during execution\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from nanocompore.Eventalign_collapse import Eventalign_collapse\n",
    "from nanocompore.common import jhelp\n",
    "jhelp(Eventalign_collapse.__init__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CLI documentation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "nanocompore eventalign_collapse --help\n",
    "```\n",
    "\n",
    "```text\n",
    "usage: nanocompore eventalign_collapse [-h] [--eventalign EVENTALIGN]\n",
    "                                       [--n_lines N_LINES]\n",
    "                                       [--nthreads NTHREADS]\n",
    "                                       [--outpath OUTPATH]\n",
    "                                       [--outprefix OUTPREFIX] [--overwrite]\n",
    "                                       [--log_level {warning,info,debug}]\n",
    "                                       [--progress]\n",
    "\n",
    "Collapse the nanopolish eventalign output at kmers level and compute kmer level statistics\n",
    "* Minimal example\n",
    "    nanocompore eventalign_collapse -i nanopolish_eventalign.tsv -outprefix out\n",
    "\n",
    "optional arguments:\n",
    "  -h, --help            show this help message and exit\n",
    "\n",
    "Input options:\n",
    "  --eventalign EVENTALIGN, -i EVENTALIGN\n",
    "                        Path to a nanopolish eventalign tsv output file, or a\n",
    "                        list of file, or a regex (can be gzipped). It can be\n",
    "                        ommited if piped to standard input (default: piped to\n",
    "                        stdin)\n",
    "\n",
    "Run parameters options:\n",
    "  --n_lines N_LINES, -l N_LINES\n",
    "                        Number of lines to parse.(default: no limits\n",
    "\n",
    "Other options:\n",
    "  --nthreads NTHREADS, -t NTHREADS\n",
    "                        Total number of threads. 2 threads are reserved for\n",
    "                        the reader and the writer (default: 3)\n",
    "\n",
    "Output options:\n",
    "  --outpath OUTPATH, -o OUTPATH\n",
    "                        Path to the output folder (default: ./)\n",
    "  --outprefix OUTPREFIX, -p OUTPREFIX\n",
    "                        text outprefix for all the files generated (default:\n",
    "                        out)\n",
    "  --overwrite, -w       Use --outpath even if it exists already (default:\n",
    "                        False)\n",
    "\n",
    "Verbosity options:\n",
    "  --log_level {warning,info,debug}\n",
    "                        Set the log level (default: info)\n",
    "  --progress            Display a progress bar during execution (default:\n",
    "                        False)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-08T11:54:48.325509Z",
     "start_time": "2020-12-08T11:54:48.314150Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(range(0,1))"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python [conda env:NanoCompore]",
   "language": "python",
   "name": "conda-env-NanoCompore-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
